rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+tied to a particular higher level framework (e.g. Keras, Estimator or Sonnet).
+The module abstraction standardizes naming of variables and ops (to simplify
+debugging/visualisation) and dependency tracking (e.g. retrieving `tf.Variable`s
+and `tf.Module`s referenced by a given root module).
+
+Having a standard abstraction for stateful TensorFlow components will encourage
+users to create loosely coupled components which are easy to reason about in
+isolation while providing a standard mechanism for common tasks in a larger
+assembly (e.g. finding all trainable variables for a module tree given a root).
+
+We see such a lightweight abstraction as being particularly important for
+cutting edge research, where it is not possible for library/framework authors to
+reason about what their users will create. We believe that lightweight
+abstractions which can be extended or composed to build more complex systems are
+the way forward.
+
+## Background
+
+TensorFlow V1 has a number of abstractions for handling TensorFlow state (e.g.
+`tf.Variable`s). Broadly you can split them into "collections based" (e.g.
+`tf.variable_scope`, `tf.get_collection`, `tf.make_template`, `tf.layers.Layer`)
+and "Object Oriented" (e.g. `tf.keras.layers.Layer`).
+
+TensorFlow V2 is standardising on OO and `tf.keras` as the only stateful
+abstraction inside the default TensorFlow pip installation.
+
+Outside of TensorFlow there exist a number of third party libraries which also
+offer abstractions for managing TensorFlow state. For example the Sonnet library
+from DeepMind is built on top of `tf.make_template`.
+
+## Requirements
+
+### Functional
+
+**Naming `tf.Variable`s (and `tf.Operation`s)**
+
+-   Each module instance must have a name scope. This name should include the
+    name scope from when the module is constructed.
+-   Every method on the module must enter this name scope before calling user
+    code (such that variables and operations are correctly named).
+
+**`tf.Variable` creation and reuse**
+
+-   Modules should be able to create and reuse variables.
+-   Users should be able to intercept variable creation without changing module
+    code (e.g. via `tf.variable_creator_scope`).
+-   Users should be able to access and change variables on a module in an
+    intuitive way (e.g. being able to copy variables between modules
+    `module.w = other_module.w`).
+
+**`tf.Variable` tracking**
+
+-   Variables created, owned or used by a module should be retrievable (e.g.
+    mod.variables should provide a collection of variables for mod and all of
+    its descendent modules).
+-   Which variables are tracked should be obvious (e.g. all object properties)
+    and "opt-out" rather than "opt-in" (e.g. the most obvious way to create and
+    reuse a variable should automatically be tracked).
+
+**`tf.Module` tracking**
+
+-   Modules should track dependencies (e.g. a sequential should provide access
```

Comment:
Something that started on twitter -> https://twitter.com/lc0d3r/status/1090351739537145858

Since we are trying to also 
```
track dependencies (e.g. a sequential should provide access to its layers) and support checkpointing as a graph.
```

Would it make sense to think about the next step and have on the model level an approachable way to have access to DAG.

The idea is, once you inherit from tf.Module you can run something similar to `keras` `model.summary`  and even more exciting - `model.plot_model`.

Sorry for a bit confusing way to formulate this suggestion

Reply:
Hey @lc0, thanks for the suggestion! We've been thinking about something similar. I'm not sure if we should solve this inside tf.Module or if this is something we should leave to framework authors, what do you think?

FWIW the current module APIs expose a lot of useful information to do a modest job of this outside of Module:

```python
import tabulate

def weight_size(m):
  size = 0
  for v in m.variables:
    size += v.numpy().nbytes

  suffix = 'B'
  suffixes = ['KB', 'MB', 'GB', 'TB']
  while size > 1024 and suffixes:
    size /= 1024
    suffix = suffixes.pop(0)
  return '%d %s' % (size, suffix)

def summarise(module):
  rows = []
  for m in module.submodules:
    row = [m.name, m.__class__.__name__, weight_size(m)]
    rows.append(row)
  print(tabulate.tabulate(rows, headers=['attr', 'class', 'size'], tablefmt='grid'))
```

```
+----------+---------+--------+
| attr     | class   | size   |
+==========+=========+========+
| linear_0 | Linear  | 8 KB   |
+----------+---------+--------+
| linear_1 | Linear  | 4 MB   |
+----------+---------+--------+
| linear_2 | Linear  | 4 MB   |
+----------+---------+--------+
| linear_3 | Linear  | 40 KB  |
+----------+---------+--------+
```

Reply:
The table seems excessive. Here is something with less chrome:

```
 attr      class    size   
+========+=======+======+
 linear_0  Linear   8 KB   
 linear_1  Linear   4 MB   
 linear_2  Linear   4 MB   
 linear_3  Linear   40 KB  
```